### PR TITLE
Doc sketch

### PR DESCRIPTION
**Pull request by rikvl**
Hi Marten, This is not ready yet. I made a 'draft pull request' so you can review my work.

### CODE REVIEW

File: docs/example_figure.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    # Create dynamic wavefield and dynamic spectrum
+    th_perp = np.zeros_like(th_par)
+    dynwave = dynamic_field(th_par, th_perp, magnifs, d_eff, mu_eff, f, t)
+    axes = tuple(range(0, dynwave.ndim-2))
+    dynwave = dynwave[...].sum(axes)
+    dynspec = np.abs(dynwave)**2
+
+    # Make conjugate and secondary spectrum
+    npad = 1
+    dspad = np.pad(dynspec, ((0, npad*nt), (0, npad*nf)),
```
Comment (mhvk) [inline docs/example_figure.py:40]: Why is this done? If the odd number in `nt` and `nf` is annoying, perhaps change it to even and use `endpoint=False` in the `np.linspace` calls.

File: docs/example_figure.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                   mode='constant', constant_values=np.mean(dynspec))
+    conspec = np.fft.fft2(dspad)
+    conspec /= conspec[0, 0]
+    tau = np.fft.fftfreq(dspad.shape[1], f[1] - f[0]).to(u.us)
+    fd = np.fft.fftfreq(dspad.shape[0], t[1] - t[0]).to(u.mHz)
+
+    conspec = np.fft.fftshift(conspec)
+    tau = np.fft.fftshift(tau) << tau.unit
```
Comment (mhvk) [inline docs/example_figure.py:48]: The `<< tau.unit` shouldn't be necessary any more (it was for astropy < 4.0); same for next line

File: docs/example_figure.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    t = np.linspace(-0.5*delta_t, 0.5*delta_t, nt)[:, np.newaxis]
+
+    # Create dynamic wavefield and dynamic spectrum
+    th_perp = np.zeros_like(th_par)
+    dynwave = dynamic_field(th_par, th_perp, magnifs, d_eff, mu_eff, f, t)
+    axes = tuple(range(0, dynwave.ndim-2))
+    dynwave = dynwave[...].sum(axes)
+    dynspec = np.abs(dynwave)**2
+
+    # Make conjugate and secondary spectrum
```
Comment (mhvk) [inline docs/example_figure.py:38]: More general: I wonder if it would make sense to just use `SecSpec.from_dynspec`?

File: docs/example_figure.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    cbar = plt.colorbar(im1)
+    cbar.set_label('normalized intensity')
+
+    # Plot secondary spectrum
+    ax2 = plt.subplot(223)
+    ss_extent = (fd[0].value, fd[-1].value, tau[0].value, tau[-1].value)
+    im2 = ax2.imshow(secspec.T, origin='lower', aspect='auto',
+                     extent=ss_extent, cmap='Greys',
+                     norm=colors.LogNorm(vmin=1.e-7, vmax=1.))
+    plt.xlabel('differential Doppler shift $f_\\mathrm{D}$ [%s]'
```
Comment (mhvk) [inline docs/example_figure.py:73]: A bit more modern would be `rf"differential Doppler shift $f_{D}$ ({fd.unit.to_string('latex')})"`
(or use `.format` instead of `%`)

File: docs/example_figure.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
+from matplotlib import pyplot as plt
```
Comment (mhvk) [inline docs/example_figure.py:1]: Should we put this file under "examples"?

File: docs/simple_examples.rst [hunk 6]
```diff
@@ -0,0 +1,148 @@
+******************************
+Simple scintillometry examples
+******************************
+
+Consider observing a pulsar using a single radio dish.
+Let's assume that the pular is stationary w.r.t. Earth,
```
Comment (mhvk) [inline docs/simple_examples.rst:6]: I'd do "Let's first assume " and omit "that its spectrum is flat, OTHER" - or, rather, move them to the end in the notes section, adding a sentence about "implicit assumptions".

Comment (mhvk) [inline docs/simple_examples.rst:6]: Nitpick: avoid abbreviations in documentation, so just "with respect to" (or just omit it, pulsar stationary is clear enough).

File: docs/simple_examples.rst [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+    theta = [-2., 0.] << u.mas
+    magnifs = gen_magnifs(theta)
+
+    example_figs(theta, magnifs)
+
+Two beams
+=========
+
+An single additional scattering image in the lens will result in
```
Comment (mhvk) [inline docs/simple_examples.rst:72]: An -> a

File: docs/simple_examples.rst [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+.. plot::
+   :context: reset
+
+    from astropy import units as u
```
Comment (mhvk) [inline docs/simple_examples.rst:77]: Not very important, but is it possible to re-use the three import lines throughout the file? (maybe by defining a context?)

File: docs/simple_examples.rst [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        | bandpass                    | :math:`\Delta f`         | 1 MHz     |
+        +-----------------------------+--------------------------+-----------+
+        | observation length          | :math:`\Delta t`         | 20 min    |
+        +-----------------------------+--------------------------+-----------+
+        | frequency channels          | :math:`n_f`              | 201       |
+        +-----------------------------+--------------------------+-----------+
+        | time bins                   | :math:`n_t`              | 101       |
+        +-----------------------------+--------------------------+-----------+
+
+Scaling with distance and observing frequency
```
Comment (mhvk) [inline docs/simple_examples.rst:145]: Let's postpone this for follow-up.

File: docs/sketches.rst [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+Making sketches
+***************
+
+
+
+Generating a schematic
+======================
+
+Consider a 1-dimensional scattering system, with scattering points located at
+angles ``theta`` w.r.t. the direct line of light (in the direction parallel to
```
Comment (mhvk) [inline docs/sketches.rst:11]: Really parallel to the effective proper motion.

File: docs/sketches.rst [hunk 11]
```diff
@@ -0,0 +1,48 @@
+***************
```
Comment (mhvk) [inline docs/sketches.rst:1]: I'd be inclined to move this to another PR. I'm a bit worried about setting the `make_sketch` API in stone...

**Review by mhvk — COMMENTED** [review]
Looking very nice!

File: docs/example_figure.py (hunk 2 above)
Comment (rikvl) [reply docs/example_figure.py:48]: Removed

File: docs/example_figure.py (hunk 1 above)
Comment (rikvl) [reply docs/example_figure.py:40]: The padding is just used to get a higher resolution in the secondary spectrum. I followed some of Daniel's code. We could leave it out, but the secondary spectra will look a bit grainy and the "flare" seems to be worse.

Comment (mhvk) [reply docs/example_figure.py:40]: Let's leave it out. Note that it is fine to make the dynamic spectrum bigger and/or have finer resolution.

File: docs/example_figure.py (hunk 3 above)
Comment (rikvl) [reply docs/example_figure.py:38]: I just tried this and the resulting secondary spectrum seems much brighter (more power in the points other than the origin). Is this secondary spectrum normalized to the power at the origin? Maybe I did something very wrong when producing the conjugate / secondary spectrum.

Comment (mhvk) [reply docs/example_figure.py:38]: Yes, it should be normalized at the origin. But did you at the same time remove the padding?  That padding does reduce the brightness (since you just add a lot of pixels with just mean value).

File: docs/conf.py [hunk 12]
```diff
@@ -43,6 +43,27 @@
...
 
+
+# -- Plot options (adapted from Astropy docs/conf.py) -------------------------
+
+plot_rcparams = {}
+# plot_rcparams['figure.figsize'] = (6, 6)
+plot_rcparams['savefig.facecolor'] = 'none'
+plot_rcparams['savefig.bbox'] = 'tight'
+plot_rcparams['axes.labelsize'] = 'large'
+plot_rcparams['figure.subplot.wspace'] = 0.0
```
[the remark this answers is not in the document]
Comment (rikvl) [reply docs/conf.py:54]: Done.

File: docs/example_figure.py (hunk 1 above)
Comment (rikvl) [reply docs/example_figure.py:40]: Irrelevant now -- see reply to your next comment about using `SecSpec.from_dynspec`

File: docs/example_figure.py (hunk 3 above)
Comment (rikvl) [reply docs/example_figure.py:38]: I got confused because `SecondarySpectrum.secspec` seems to be (what I thought is referred to as) the conjugate spectrum.

File: docs/example_figure.py (hunk 4 above)
Comment (rikvl) [reply docs/example_figure.py:73]: Done, but I kept the `\mathrm`, because I thought the subscript D is not a physical parameter, but an abbreviation for Doppler.

File: docs/example_figure.py (hunk 5 above)
Comment (rikvl) [reply docs/example_figure.py:1]: I put both this python file and the rst file in a dir called "examples".

In general, if you have good ideas about file/variable/function names and/or file structure, please implement them or let me know. I tend to use placeholder names to avoid getting blocked at thinking of the perfect name for a file.

File: docs/simple_examples.rst (hunk 6 above)
Comment (rikvl) [reply docs/simple_examples.rst:6]: Done

Comment (rikvl) [reply docs/simple_examples.rst:6]: Done

File: docs/simple_examples.rst (hunk 7 above)
Comment (rikvl) [reply docs/simple_examples.rst:72]: Gone

File: docs/simple_examples.rst (hunk 8 above)
Comment (rikvl) [reply docs/simple_examples.rst:77]: Hmmm, in the Astropy docs example I was following the imports are repeated. We could put some things in the `plot_pre_code` in `conf.py`, but to me that seems like making the code harder to debug.

File: docs/simple_examples.rst (hunk 9 above)
Comment (rikvl) [reply docs/simple_examples.rst:145]: I commented it out

File: docs/sketches.rst (hunk 11 above)
Comment (rikvl) [reply docs/sketches.rst:1]: Good idea

File: docs/sketches.rst (hunk 10 above)
Comment (rikvl) [reply docs/sketches.rst:11]: Noted

File: docs/example_figure.py (hunk 3 above)
Comment (mhvk) [reply docs/example_figure.py:38]: Hmm, yes, I think my nomenclature is bad here! Maybe we should rename the class...

File: docs/example_figure.py (hunk 4 above)
Comment (mhvk) [reply docs/example_figure.py:73]: Yes, makes sense.

File: docs/simple_examples.rst (hunk 8 above)
Comment (mhvk) [reply docs/simple_examples.rst:77]: I was hoping we could those common bits in a separate section in the same file. If that's not easily possible, then definitely just keep things as they are.

File: docs/examples/example_figure.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
+from matplotlib import pyplot as plt
+from matplotlib import colors as colors
+import numpy as np
+from astropy import units as u
+from screens.fields import dynamic_field
+from screens.secspec import SecondarySpectrum
+from screens.visualization import make_sketch
+
+
+def gen_magnifs(th):
```
Comment (mhvk) [inline docs/examples/example_figure.py:10]: Nitpick, but since in `dynspec` and `secspec`, this is called `magnification`, maybe just stick to that? Maybe `generate_magnification`

File: docs/examples/example_figure.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    plt.ylim(-35., 35.)
+    cbar = plt.colorbar(im2)
+    cbar.set_label('normalized power')
+
+    # Add sketch to the right
+    ax3 = plt.subplot(122)
+    make_sketch(th_par, rotation=-90.*u.deg, ax=ax3, **kwargs)
+    plt.title('schematic')
+
+    plt.show()
```
Comment (mhvk) [inline docs/examples/example_figure.py:83]: My sense would be to also include code to generate an actual example figure, i.e.,
```
if __name__ == '__main__':
    <code for, say, the multi-point final figure>
```

File: docs/examples/example_figure.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    a = 0.3*np.exp(-0.5*(th/sig)**2).to_value(1)
+    np.random.seed(2)
+    magnifs = a * np.random.normal(size=(th.size, 2)).view('c16').squeeze(-1)
+    magnifs[np.where(th == 0)] = 1
+    magnifs /= np.sqrt((np.abs(magnifs)**2).sum())
+
+    return magnifs
+
+
+def example_figs(th_par, magnifs,
```
Comment (mhvk) [inline docs/examples/example_figure.py:23]: Maybe just name if `make_figure`, `simple_figure` or so? We're likely to have more figures in here soon... (can also call the file `simple_figure` and call the function `figure`)

File: docs/examples/simple_examples.rst [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+A scattering screen
+===================
+
+Now introduce a scattering screen somewhere in the space between the pulsar and
+Earth. This screen, a one-dimensional lens, will give rise to point-like
+scattering images of the pulsar. As a first step, we consider the effect of a
+single scattering image. In addition to the direct line of sight, there will
+now be a scattered beam via which radiation from the pulsar reaches Earth.
+
+The radiation going via the
```
Comment (mhvk) [inline docs/examples/simple_examples.rst:33]: Super-nitpick, but if could you split sentences into lines of roughly equal lines?

File: docs/examples/simple_examples.rst [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+the effective velocity of the system.
+
+.. plot::
+
+    from astropy import units as u
+    import matplotlib.pyplot as plt
+    from example_figure import gen_magnifs, example_figs
+
+    theta = [0., 2.] << u.mas
+    magnifs = gen_magnifs(theta)
```
Comment (mhvk) [inline docs/examples/simple_examples.rst:79]: Here, using exactly the same magnifications as in the previous example would be helpful.

File: docs/examples/simple_examples.rst [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+Comparing the 
+
+.. plot::
+
+    from astropy import units as u
+    import matplotlib.pyplot as plt
+    from example_figure import gen_magnifs, example_figs
+
+    theta = [-4., 0., 2.] << u.mas
+    magnifs = gen_magnifs(theta)
```
Comment (mhvk) [inline docs/examples/simple_examples.rst:107]: Explicit magnifications here too.

File: docs/examples/simple_examples.rst [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+Adding more beams will further complicate the interference pattern
+
+.. plot::
+
+    from astropy import units as u
+    import matplotlib.pyplot as plt
+    from example_figure import gen_magnifs, example_figs
+
+    theta = [-4., -1., 0., 2.] << u.mas
+    magnifs = gen_magnifs(theta)
```
Comment (mhvk) [inline docs/examples/simple_examples.rst:123]: And here.

File: docs/examples/simple_examples.rst [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+        | bandpass                    | :math:`\Delta f`         | 1 MHz     |
+        +-----------------------------+--------------------------+-----------+
+        | observation length          | :math:`\Delta t`         | 20 min    |
+        +-----------------------------+--------------------------+-----------+
+        | frequency channels          | :math:`n_f`              | 201       |
+        +-----------------------------+--------------------------+-----------+
+        | time bins                   | :math:`n_t`              | 101       |
+        +-----------------------------+--------------------------+-----------+
+
+.. Scaling with distance and observing frequency
```
Comment (mhvk) [inline docs/examples/simple_examples.rst:171]: I'd leave this out of the text and raise an issue instead.

File: screens/visualization/sketch.py [hunk 21]
```diff
@@ -155,59 +158,72 @@ def make_sketch(theta, beta=0.5, screen_y_scale=1.e7,
...
+    for lens in lens_pos:
         p2s2t = []
         for p2t in p2t_pos:
             p = tuple(p.copy() for p in p2t)
-            p[0][1] = screen_x
-            p[1][1] = s_y
-            p[1][0] = s_y * shortening[0]/shortening[1]
+            p[0][1] = lens[0]
+            p[1][1] = lens[1]
+            p[0][0] = ns_offset[0] \
```
Comment (mhvk) [inline screens/visualization/sketch.py:206]: More nit-picking: `\` is generally discouraged. Instead, put parentheses around the expression, like
```
p[0][0] = (ns_offset[0]
           + shortening[0]/shortening[1]
           * (lens[0] - ns_offset[0]))
```

File: screens/visualization/sketch.py [hunk 22]
```diff
@@ -155,59 +158,72 @@ def make_sketch(theta, beta=0.5, screen_y_scale=1.e7,
...
     # Scattering screen
+    screen_size = 12.
     screen_pos = []
     for a, f, p in zip([0.5, 0.2, 0.3],
                        [1., 0.7, 1.6] * u.cycle,
                        [0.1, 0.4, 0.5] * u.cycle):
-        y = np.linspace(-6., 6., 361) * u.dimensionless_unscaled
-        x = screen_x + a * np.cos(f*y + p)
-        screen_pos += [x, y]
+        y = np.linspace(-screen_size/2., screen_size/2., 361) \
```
Comment (mhvk) [inline screens/visualization/sketch.py:219]: As above.

**Review by mhvk — COMMENTED** [review]
Nice! Left are nitpicks...

File: docs/examples/example_figure.py (hunk 13 above)
Comment (rikvl) [reply docs/examples/example_figure.py:10]: Done

File: docs/examples/example_figure.py (hunk 15 above)
Comment (rikvl) [reply docs/examples/example_figure.py:23]: Done. These names may have to change when we have a better (idea of) the structure of the docs.

File: docs/examples/example_figure.py (hunk 14 above)
Comment (rikvl) [reply docs/examples/example_figure.py:83]: Done.

Comment (mhvk) [reply docs/examples/example_figure.py:83]: Hmm, not actually done!

Comment (rikvl) [reply docs/examples/example_figure.py:83]: @mhvk Sorry about that Marten, I'm still getting used to the github interface.

I'd made the changes locally and wrote these comments to tick them off here, but I should only do that after pushing the changes.

The most recent commit should have implemented all your comments.

File: docs/examples/example_figure.py [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+    ax3.imshow(np.angle(dynwave).T,
+               alpha=np.abs(dynwave).T / np.max(np.abs(dynwave).T),
+               origin='lower', aspect='auto', interpolation='none',
+               cmap=phasecmap, extent=ds_extent, vmin=-np.pi, vmax=np.pi)
+    plt.title('electric field')
+    plt.xlabel(rf"time $t$ ({t.unit.to_string('latex')})")
+    plt.ylabel(rf"frequency $f$ ({f.unit.to_string('latex')})")
+    plt.xlim(t_lims)
+    plt.ylim(f_lims)
+    # cbar = plt.colorbar(im3)
```
Comment (mhvk) [inline docs/examples/example_figure.py:121]: Remove comment?

File: docs/examples/simple_examples.rst [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+When a relatively large number of beams is considered, the dynamic spectrum
+becomes so complex that the interference patterns due to individual pairs of
+beams can no longer been identified. It becomes more practical to characterize
+the dynamic spectrum as a pattern of bright (amplified) patches in time and
+frequency (known as "scintils") against a less bright background.
+
+In the secondary spectrum, the points arrising from interaction of a scattered
+image with the central one line up along parabolas with their extrema at the
+origin. The points caused by mutual interaction of scattered beam form
+"arclets", interted parabolas positioned along the main parabolas.
```
Comment (mhvk) [inline docs/examples/simple_examples.rst:148]: interted->inverted

**Review by mhvk — COMMENTED** [review]
Two very small things only. If you could address those and squash the commits, we can put this in!

File: docs/examples/example_figure.py (hunk 23 above)
Comment (rikvl) [reply docs/examples/example_figure.py]: Done

File: docs/examples/simple_examples.rst (hunk 24 above)
Comment (rikvl) [reply docs/examples/simple_examples.rst]: Done

**Comment by rikvl** [comment]
All done, so I think you can merge it now :)
